crates/compiler/build/src/program.rs

```diff
@@ -1002,36 +1064,37 @@ fn build_loaded_file<'a>(
 
             std::fs::write(app_o_file, &*roc_app_bytes).unwrap();
 
-            let builtins_host_tempfile = roc_bitcode::host_tempfile()
-                .expect("failed to write host builtins object to tempfile");
-
             let mut inputs = vec![app_o_file.to_str().unwrap()];
 
-            if !matches!(link_type, LinkType::Dylib | LinkType::None) {
-                // the host has been compiled into a .o or .obj file
-                inputs.push(preprocessed_host_path.as_path().to_str().unwrap());
+            let mut host_path = String::new();
+            if let PrebuiltHost::Legacy(p) = built_host_path {
+                host_path.push_str(&p.to_string_lossy());
+                inputs.push(&host_path);
+            } else {
+                // we may be compiling a Dylib, which doesn't have a host
+                // for example, we could be generating glue using `roc glue`
             }
 
+            let builtins_host_tempfile = roc_bitcode::host_tempfile()
+                .expect("failed to write host builtins object to tempfile");
+
             if matches!(code_gen_options.backend, CodeGenBackend::Assembly(_)) {
                 inputs.push(builtins_host_tempfile.path().to_str().unwrap());
             }
 
             let (mut child, _) = link(target, output_exe_path.clone(), &inputs, link_type)
-                .map_err(|_| todo!("gracefully handle `ld` failing to spawn."))?;
+                .map_err(|_| todo!("linker failed to spawn."))?;
 
             let exit_status = child
                 .wait()
-                .map_err(|_| todo!("gracefully handle error after `ld` spawned"))?;
+                .map_err(|_| todo!("linker error after spawning"))?;
 
             // Extend the lifetime of the tempfile so it doesn't get dropped
             // (and thus deleted) before the child process is done using it!
             let _ = builtins_host_tempfile;
 
             if !exit_status.success() {
-                todo!(
-                    "gracefully handle `ld` (or `zig` in the case of wasm with --optimize) returning exit code {:?}",
-                    exit_status.code()
-                );
+                todo!("linker failed with exit code {:?}", exit_status.code());
```

Comment:
TODO -- run through this PR and double check errors for where we might use `internal_error!` macro